File: Example/DeepAdxAdsDebug/DeepAdxDebug/Demo/ExampleController/Template/DemoListFeedExpressViewController.m
//
//  DemoListFeedExpressViewController.m
//  DeepAdxDebug
//
//  Created by cc on 2022/8/30.
//

#import "DemoListFeedExpressViewController.h"
#import <DeepAdxAdspot/DeepAdxNativeExpress.h>
#import <DeepAdxAdspot/DeepAdxNativeExpressView.h>
#import "DemoAdConfig.h"

@interface DemoListFeedExpressViewController ()<UITableViewDelegate, UITableViewDataSource, DeepAdxNativeExpressDelegate> {
    BOOL _isLoadAndShow;
    BOOL _isShowLogView;
    CGFloat _navAndStateBarHeight;
}

//@property (strong, nonatomic) UITableView *tableView;
@property (strong,nonatomic) DeepAdxNativeExpress *advanceFeed;
@property (nonatomic, strong) NSMutableArray *dataArrM;
@property (nonatomic, strong) NSMutableArray *arrViewsM;
@property (nonatomic, assign) BOOL isLoadAndShow;
@property (nonatomic, assign) BOOL isShowLogView;
/// 视图容器
@property(nonatomic, strong) UIView *containerView;
@end

@implementation DemoListFeedExpressViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.navigationItem.title = @"信息流";
    
    CGFloat statusHeight = [[UIApplication sharedApplication] statusBarFrame].size.height;
    CGFloat navHeight = self.navigationController.navigationBar.frame.size.height;
    
    _navAndStateBarHeight = navHeight + statusHeight;
    
    self.textV.frame = CGRectMake(0,0 - 300, self.view.frame.size.width, 300);

    UIBarButtonItem *rightBar = [[UIBarButtonItem alloc] initWithTitle:@"显示log信息" style:UIBarButtonItemStylePlain target:self action:@selector(showLogView)];
    self.navigationItem.rightBarButtonItem = rightBar;
}

- (void)loadAd {
    [super loadAd];
    [self deallocAd];
    [self loadAdWithState:AdState_Normal];
    _isLoadAndShow = NO;
    self.dataArrM = [NSMutableArray array];
    [self.advanceFeed loadAd];
    [self loadAdWithState:AdState_Loading];
}

- (void)showAd {
    if (!self.advanceFeed || !self.isLoaded || self.arrViewsM.count == 0) {
        return;
    }
    [self showNativeAd];
}

- (void)loadAndShowAd {
    [super loadAd];
    [self deallocAd];
    [self loadAdWithState:AdState_Normal];
    _isLoadAndShow = YES;
    self.dataArrM = [NSMutableArray array];
    [self.advanceFeed loadAndShowAd];
    [self loadAdWithState:AdState_Loading];
}

- (void)deallocAd {
    if (_advanceFeed) {
        _advanceFeed.delegate = nil;
        _advanceFeed = nil;
    }
    self.isLoaded = NO;
    [self.dataArrM removeAllObjects];
    [self.arrViewsM removeAllObjects];
    [self loadAdWithState:AdState_Normal];
}

// 信息流广告比较特殊, 渲染逻辑需要自行处理
- (void)showNativeAd {
    for (NSInteger i = 0; i < self.arrViewsM.count; i++) {
        DeepAdxNativeExpressView *view = self.arrViewsM[i];
        [view render];
        [_dataArrM addObject:self.arrViewsM[i]];
    }
    
    if (_dataArrM.count <= 0) {
        return;
    }
    
    [_containerView removeFromSuperview];
    UIView *view = [_dataArrM.firstObject expressView];
    self.containerView = view;
    [self.view addSubview:self.containerView];

    [self.containerView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.right.mas_equalTo(0);
        make.top.mas_equalTo(330);
        // 占位高度：约束高度过小会导致 hitTest 区域几乎为 0，表现为「看得见点不了」；回调里会更新为真实值
        make.height.mas_equalTo(300);
    }];
    [self.view bringSubviewToFront:self.containerView];
}

- (void)showLogView {
    [UIView animateWithDuration:0.2 animations:^{
        self.textV.frame = CGRectMake(0,((self->_isShowLogView = !self->_isShowLogView) ? self->_navAndStateBarHeight : -300 ), self.view.frame.size.width, 300);
        self.navigationItem.rightBarButtonItem.title = self->_isShowLogView ? @"隐藏log信息":@"显示log信息";
    }];
}

#pragma mark - DeepAdxNativeExpressDelegate

/// 内部渠道开始加载时调用
- (void)ad_supplierWillLoad:(NSString *)supplierId {
    NSLog(@"内部渠道开始加载 %s  supplierId: %@", __func__, supplierId);
    [self showProcessWithText:[NSString stringWithFormat:@"%s\r\n 内部渠道开始加载", __func__]];
}

/// 选中渠道
- (void)ad_successSortTag:(NSString *)sortTag {
    NSLog(@"选中了 rule '%@' %s", sortTag,__func__);
    [self showProcessWithText:[NSString stringWithFormat:@"%s\r\n 选中了 rule '%@' ", __func__, sortTag]];
}

/// 广告数据拉取成功
- (void)ad_NativeExpressOnAdLoadSuccess:(nullable NSArray<DeepAdxNativeExpressView *> *)views {
    NSLog(@"广告拉取成功 %s", __func__);
    self.arrViewsM = [NSMutableArray arrayWithArray:views];
    
    if (_isLoadAndShow) {
        [self showNativeAd];
    }
    
    self.isLoaded = YES;
    [self showProcessWithText:[NSString stringWithFormat:@"%s\r\n 广告拉取成功", __func__]];
    [self loadAdWithState:AdState_LoadSucceed];
}

/// 广告数据拉取失败
- (void)ad_NativeExpressOnAdLoadFailWithError:(NSError *_Nullable)error {
    [self showProcessWithText:[NSString stringWithFormat:@"%s\r\n 广告加载失败", __func__]];
    [self loadAdWithState:AdState_LoadFailed];
    [self deallocAd];
}
/// 广告渲染成功
- (void)ad_NativeExpressOnAdRenderSuccess:(nullable DeepAdxNativeExpressView *)adView {
    NSLog(@"广告渲染成功 %s %@", __func__, adView);
    UIView *ev = adView.expressView;
    [ev layoutIfNeeded];
    CGFloat h = CGRectGetHeight(ev.bounds);
    if (h < 1.f) {
        h = CGRectGetHeight(ev.frame);
    }
    if (h < 1.f) {
        CGSize fit = [ev systemLayoutSizeFittingSize:UILayoutFittingCompressedSize];
        h = fit.height;
    }
    if (h < 1.f) {
        h = 300.f;
    }
    [self.containerView mas_updateConstraints:^(MASConstraintMaker *make) {
        make.height.mas_equalTo(h);
    }];
    [self showProcessWithText:[NSString stringWithFormat:@"%s\r\n 广告渲染成功", __func__]];
}
/// 广告渲染失败
- (void)ad_NativeExpressOnAdRenderFail:(nullable DeepAdxNativeExpressView *)adView withError:(NSError *_Nullable)error {
    NSLog(@"广告渲染失败 %s %@", __func__, adView);
    [self showProcessWithText:[NSString stringWithFormat:@"%s\r\n 广告渲染失败", __func__]];
    [_dataArrM removeObject: adView];
}
/// 广告视图为空
- (void)ad_NativeExpressOnAdGetViewIsEmpty {
    NSLog(@"广告视图为空");
}
/// 广告曝光
- (void)ad_NativeExpressOnAdShow:(nullable DeepAdxNativeExpressView *)adView {
    NSLog(@"广告曝光 %s", __func__);
    UIView *ev = adView.expressView;
    [ev layoutIfNeeded];
    CGFloat h = CGRectGetHeight(ev.bounds);
    if (h < 1.f) {
        h = CGRectGetHeight(ev.frame);
    }
    if (h < 1.f) {
        CGSize fit = [ev systemLayoutSizeFittingSize:UILayoutFittingCompressedSize];
        h = fit.height;
    }
    if (h < 1.f) {
        h = 300.f;
    }
    [self.containerView mas_updateConstraints:^(MASConstraintMaker *make) {
        make.height.mas_equalTo(h);
    }];
    [self showProcessWithText:[NSString stringWithFormat:@"%s\r\n 广告曝光成功", __func__]];
}
/// 广告点击
- (void)ad_NativeExpressOnAdClicked:(nullable DeepAdxNativeExpressView *)adView {
    NSLog(@"广告点击 %s", __func__);
    [self showProcessWithText:[NSString stringWithFormat:@"%s\r\n 广告点击", __func__]];
}
/// 广告被关闭 (注: 百度广告(百青藤), 不支持该回调, 若使用百青藤,则该回到功能请自行实现)
- (void)ad_NativeExpressOnAdClosed:(nullable DeepAdxNativeExpressView *)adView {
    NSLog(@"广告关闭 %s", __func__);
    [self deepadx_removeNativeExpressViewFromScreen:adView];
}

- (void)ad_NativeExpressOnAdDislike:(nullable DeepAdxNativeExpressView *)adView {
    NSLog(@"广告不喜欢 %s", __func__);
    // 部分渠道（如快手）点「不喜欢」只走 dislike，不走 closed，需主动移除 expressView
    [self deepadx_removeNativeExpressViewFromScreen:adView];
}

- (void)deepadx_removeNativeExpressViewFromScreen:(nullable DeepAdxNativeExpressView *)adView {
    if (!adView) {
        return;
    }
    [_dataArrM removeObject:adView];
    [_arrViewsM removeObject:adView];
    UIView *ev = adView.expressView;
    if (ev.superview) {
        [ev removeFromSuperview];
    }
    if (self.containerView == ev) {
        self.containerView = nil;
    }
}


#pragma mark - lazy
- (DeepAdxNativeExpress *)advanceFeed{
    if(!_advanceFeed){
        NSString *placementId = [DemoAdConfig placementIdForKind:DemoAdPlacementKindNative];
        _advanceFeed = [[DeepAdxNativeExpress alloc] initWithViewController:self
                                                                    adSize:CGSizeMake(self.view.bounds.size.width, 0)
                                                               placementId:placementId];
        _advanceFeed.delegate = self;
    }
    return _advanceFeed;
}

@end

File: Example/DeepAdxAdsSwift/DeepAdxSwift/Demo/DemoListFeedExpressViewController.swift
//
//  DemoListFeedExpressViewController.swift
//  DeepAdxSwift
//

import UIKit
import DeepAdxAdspot
import Masonry

final class DemoListFeedExpressViewController: BaseDemoViewController, DeepAdxNativeExpressDelegate {

    private var advanceFeed: DeepAdxNativeExpress?
    private var dataArrM: [DeepAdxNativeExpressView] = []
    private var arrViewsM: [DeepAdxNativeExpressView] = []
    private var isLoadAndShow = false
    private var isShowLogView = false
    private var navAndStateBarHeight: CGFloat = 0
    private var containerView: UIView?
    private let logPanelHeight: CGFloat = 300

    override func viewDidLoad() {
        super.viewDidLoad()
        navigationItem.title = "信息流"

        let statusHeight: CGFloat
        if #available(iOS 13.0, *) {
            statusHeight = view.window?.windowScene?.statusBarManager?.statusBarFrame.height ?? 0
        } else {
            statusHeight = UIApplication.shared.statusBarFrame.height
        }
        let navHeight = navigationController?.navigationBar.frame.height ?? 44
        navAndStateBarHeight = navHeight + statusHeight

        // 默认隐藏，点击右上角按钮后以下拉面板方式展示日志
        layoutLogPanel(animated: false)

        navigationItem.rightBarButtonItem = UIBarButtonItem(
            title: "显示log信息",
            style: .plain,
            target: self,
            action: #selector(showLogView)
        )
    }

    override func loadAd() {
        super.loadAd()
        deallocAd()
        loadAdWithState(.normal)
        isLoadAndShow = false
        dataArrM = []
        ensureFeed().loadAd()
        loadAdWithState(.loading)
    }

    override func showAd() {
        guard advanceFeed != nil, isLoaded, !arrViewsM.isEmpty else { return }
        showNativeAd()
    }

    override func loadAndShowAd() {
        super.loadAd()
        deallocAd()
        loadAdWithState(.normal)
        isLoadAndShow = true
        dataArrM = []
        ensureFeed().loadAndShowAd()
        loadAdWithState(.loading)
    }

    override func deallocAd() {
        advanceFeed?.delegate = nil
        advanceFeed = nil
        isLoaded = false
        dataArrM.removeAll()
        arrViewsM.removeAll()
        loadAdWithState(.normal)
    }

    private func ensureFeed() -> DeepAdxNativeExpress {
        if advanceFeed == nil {
            let f = DeepAdxNativeExpress(
                viewController: self,
                adSize: CGSize(width: view.bounds.size.width, height: 0)
            )
            f.delegate = self
            advanceFeed = f
        }
        return advanceFeed!
    }

    private func showNativeAd() {
        dataArrM = []
        for v in arrViewsM {
            v.render()
            dataArrM.append(v)
        }
        guard !dataArrM.isEmpty, let first = dataArrM.first else { return }
        let express = first.expressView
        containerView?.removeFromSuperview()
        containerView = express
        view.addSubview(express)
        express.mas_remakeConstraints { make in
            make?.left.right().mas_equalTo()(0)
            make?.top.mas_equalTo()(330)
            // 高度过小会导致点击/关闭命中区域异常；占位高度在渲染/曝光回调中更新
            make?.height.mas_equalTo()(300)
        }
        view.bringSubviewToFront(express)
        view.layoutIfNeeded()
    }

    @objc private func showLogView() {
        isShowLogView.toggle()
        layoutLogPanel(animated: true)
    }

    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        // BaseDemoViewController 会重置 textV frame，这里回写信息流页面的下拉面板布局。
        layoutLogPanel(animated: false)
    }

    private func layoutLogPanel(animated: Bool) {
        let update = {
            self.navAndStateBarHeight = self.currentNavAndStatusHeight()
            let y: CGFloat = self.isShowLogView ? self.navAndStateBarHeight : -self.logPanelHeight
            self.textV.frame = CGRect(x: 0, y: y, width: self.view.bounds.width, height: self.logPanelHeight)
            self.view.bringSubviewToFront(self.textV)
            self.navigationItem.rightBarButtonItem?.title = self.isShowLogView ? "隐藏log信息" : "显示log信息"
        }
        if animated {
            UIView.animate(withDuration: 0.2, animations: update)
        } else {
            update()
        }
    }

    private func currentNavAndStatusHeight() -> CGFloat {
        let navHeight = navigationController?.navigationBar.frame.height ?? 44
        let statusHeight: CGFloat
        if #available(iOS 13.0, *) {
            statusHeight = view.window?.windowScene?.statusBarManager?.statusBarFrame.height
                ?? UIApplication.shared.connectedScenes
                    .compactMap { $0 as? UIWindowScene }
                    .first?
                    .statusBarManager?
                    .statusBarFrame.height
                ?? 0
        } else {
            statusHeight = UIApplication.shared.statusBarFrame.height
        }
        return navHeight + statusHeight
    }

    // MARK: - DeepAdxNativeExpressDelegate

    @objc(ad_supplierWillLoad:)
    func ad_supplierWillLoad(_ supplierId: String) {
        showProcessWithText("\(#function)\r\n 内部渠道开始加载")
    }

    @objc(ad_successSortTag:)
    func ad_successSortTag(_ sortTag: String) {
        showProcessWithText("\(#function)\r\n 选中了 rule '\(sortTag)' ")
    }

    @objc(ad_NativeExpressOnAdLoadSuccess:)
    func ad_NativeExpress(onAdLoadSuccess views: [DeepAdxNativeExpressView]?) {
        guard let views = views, !views.isEmpty else { return }
        arrViewsM = views
        if isLoadAndShow {
            showNativeAd()
        }
        isLoaded = true
        showProcessWithText("\(#function)\r\n 广告拉取成功")
        loadAdWithState(.loadSucceed)
    }

    @objc(ad_NativeExpressOnAdLoadFailWithError:)
    func ad_NativeExpress(onAdLoadFailWithError error: Error?) {
        showProcessWithText("\(#function)\r\n 广告加载失败")
        loadAdWithState(.loadFailed)
        deallocAd()
    }

    @objc(ad_NativeExpressOnAdRenderSuccess:)
    func ad_NativeExpress(onAdRenderSuccess adView: DeepAdxNativeExpressView?) {
        guard let expressView = adView?.expressView else { return }
        expressView.setNeedsLayout()
        expressView.layoutIfNeeded()
        var h = expressView.bounds.size.height
        if h <= 0 {
            h = expressView.frame.size.height
        }
        if h <= 0 {
            let size = expressView.systemLayoutSizeFitting(
                CGSize(width: view.bounds.width, height: UIView.layoutFittingExpandedSize.height),
                withHorizontalFittingPriority: .required,
                verticalFittingPriority: .fittingSizeLevel
            )
            h = size.height
        }
        if h <= 0 { h = 300 }
        containerView?.mas_updateConstraints { make in
            make?.height.mas_equalTo()(h)
        }
        showProcessWithText("\(#function)\r\n 广告渲染成功")
    }

    @objc(ad_NativeExpressOnAdRenderFail:withError:)
    func ad_NativeExpress(onAdRenderFail adView: DeepAdxNativeExpressView?, withError error: Error?) {
        showProcessWithText("\(#function)\r\n 广告渲染失败")
        if let adView = adView {
            dataArrM.removeAll { $0 === adView }
        }
    }

    func ad_NativeExpressOnAdGetViewIsEmpty() {}

    @objc(ad_NativeExpressOnAdShow:)
    func ad_NativeExpress(onAdShow adView: DeepAdxNativeExpressView?) {
        guard let expressView = adView?.expressView else { return }
        expressView.setNeedsLayout()
        expressView.layoutIfNeeded()
        var h = expressView.bounds.size.height
        if h <= 0 {
            h = expressView.frame.size.height
        }
        if h <= 0 {
            let size = expressView.systemLayoutSizeFitting(
                CGSize(width: view.bounds.width, height: UIView.layoutFittingExpandedSize.height),
                withHorizontalFittingPriority: .required,
                verticalFittingPriority: .fittingSizeLevel
            )
            h = size.height
        }
        if h <= 0 { h = 300 }
        containerView?.mas_updateConstraints { make in
            make?.height.mas_equalTo()(h)
        }
        showProcessWithText("\(#function)\r\n 广告曝光成功")
    }

    @objc(ad_NativeExpressOnAdClicked:)
    func ad_NativeExpress(onAdClicked adView: DeepAdxNativeExpressView?) {
        showProcessWithText("\(#function)\r\n 广告点击")
    }

    @objc(ad_NativeExpressOnAdClosed:)
    func ad_NativeExpress(onAdClosed adView: DeepAdxNativeExpressView?) {
        deepadx_removeNativeExpressFromScreen(adView)
    }

    @objc(ad_NativeExpressOnAdDislike:)
    func ad_NativeExpress(onDislike adView: DeepAdxNativeExpressView?) {
        // 部分渠道（如快手）点「不喜欢」只走 dislike，不走 closed，需主动移除 expressView
        deepadx_removeNativeExpressFromScreen(adView)
    }

    private func deepadx_removeNativeExpressFromScreen(_ adView: DeepAdxNativeExpressView?) {
        guard let adView = adView else { return }
        dataArrM.removeAll { $0 === adView }
        arrViewsM.removeAll { $0 === adView }
        let ev = adView.expressView
        ev.removeFromSuperview()
        if containerView === ev {
            containerView = nil
        }
    }
}
